### PR TITLE
[devel] single host agency can be logging too early

### PR DESCRIPTION
**Pull request by kvahed**

### CODE REVIEW

**Comment by kvahed** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/7765/

**Comment by kvahed** [comment]
Tests fine apart from VS errors:
https://jenkins01.arangodb.biz/job/arangodb-matrix-pr-windows/1738/

File: arangod/Agency/Node.cpp [hunk 1]
```diff
@@ -522,15 +522,15 @@ ResultT<std::shared_ptr<Node>> Node::handle<INCREMENT>(VPackSlice const& slice)
 /// Decrement integer value or set -1
 template <>
 ResultT<std::shared_ptr<Node>> Node::handle<DECREMENT>(VPackSlice const& slice) {
-  int inc = (slice.hasKey("step") && slice.get("step").isUInt())
+  int64_t inc = (slice.hasKey("step") && slice.get("step").isUInt())
```
Comment (jsteemann) [inline arangod/Agency/Node.cpp]:
```suggestion
  int64_t inc = (slice.hasKey("step") && slice.get("step").isInt())
```

File: arangod/Agency/State.cpp [hunk 2]
```diff
@@ -95,8 +95,16 @@ inline static std::string stringify(index_t index) {
 bool State::persist(index_t index, term_t term, uint64_t millis,
                     arangodb::velocypack::Slice const& entry,
                     std::string const& clientId) const {
-  LOG_TOPIC("b735e", TRACE, Logger::AGENCY) << "persist index=" << index << " term=" << term
-                                   << " entry: " << entry.toJson();
+  LOG_TOPIC("b735e", TRACE, Logger::AGENCY)
+    << "persist index=" << index << " term=" << term << " entry: " << entry.toJson();
+
+#ifdef MAINTAINER_MODE // we should never get here
```
Comment (jsteemann) [inline arangod/Agency/State.cpp]:
```suggestion
#ifdef ARANGODB_USE_MAINTAINER_MODE // we should never get here
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8343/

File: arangod/Agency/State.cpp [hunk 3]
```diff
@@ -95,8 +95,16 @@ inline static std::string stringify(index_t index) {
 bool State::persist(index_t index, term_t term, uint64_t millis,
                     arangodb::velocypack::Slice const& entry,
                     std::string const& clientId) const {
-  LOG_TOPIC("b735e", TRACE, Logger::AGENCY) << "persist index=" << index << " term=" << term
-                                   << " entry: " << entry.toJson();
+  LOG_TOPIC("b735e", TRACE, Logger::AGENCY)
+    << "persist index=" << index << " term=" << term << " entry: " << entry.toJson();
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE // we should never get here
+  if (index == 1 && entry.length != 1 && !entry.hasKey(RECONFIGURATION)) {
```
Comment (jsteemann) [inline arangod/Agency/State.cpp]:
```suggestion
  if (index == 1 && entry.length() != 1 && !entry.hasKey(RECONFIGURATION)) {
```

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8347/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/8352/

**Comment by jsteemann** [comment]
Jenkins tests fail now because agency can't be started, e.g.
http://172.16.10.101:8080/job/arangodb-matrix-pr-windows/2322/EDITION=enterprise,STORAGE_ENGINE=rocksdb,TEST_SUITE=single,limit=windows&&test/artifact/testfailures.log
``` 
* Test "active_failover"
    [FAILED]  setup

      "test" failed: failed to start server!


   Suites failed: 1 Tests Failed: 0


--------------------------------------------------------------------------------
Crash analysis of: {"role":"agent","port":"13105","endpoint":"tcp://127.0.0.1:13105","rootDir":"c:\\vm03-windows\\oskar\\work\\tmp\\arangosh_a07908\\client_resilience\\agency","url":"http://127.0.0.1:13105","args":{"configuration":"c:\\vm03-windows\\oskar\\work\\ArangoDB\\etc\\testing\\arangod-agency.conf","define":"TOP_DIR=c:\\vm03-windows\\oskar\\work\\ArangoDB","wal.flush-timeout":30000,"javascript.app-path":"c:\\vm03-windows\\oskar\\work\\tmp\\arangosh_a07908\\client_resilience\\agency\\app0","javascript.copy-installation":false,"http.trusted-origin":"all","cluster.create-waits-for-sync-replication":false,"temp.path":"c:\\vm03-windows\\oskar\\work\\tmp\\arangosh_a07908\\client_resilience\\agency\\tmp","server.storage-engine":"rocksdb","server.endpoint":"tcp://127.0.0.1:13105","database.directory":"c:\\vm03-windows\\oskar\\work\\tmp\\arangosh_a07908\\client_resilience\\agency\\data0","log.file":"c:\\vm03-windows\\oskar\\work\\tmp\\arangosh_a07908\\client_resilience\\agency\\log0","server.authentication":"true","server.jwt-secret":"haxxmann","javascript.allow-external-process-control":"true","javascript.allow-port-testing":"true","javascript.enabled":"false","agency.activate":"true","agency.size":"3","agency.pool-size":"3","agency.wait-for-sync":"false","agency.supervision":"true","agency.my-address":"tcp://127.0.0.1:13105","agency.supervision-grace-period":"10.0","agency.supervision-frequency":"1.0"},"pid":10740,"name":"agent - 13105","coreFilePattern":"c:\\jenkins\\workspace\\arangodb-matr---490f9675\\EDITION\\enterprise\\STORAGE_ENGINE\\rocksdb\\TEST_SUITE\\single\\limit\\windows&&test\\core\\arangod.exe.10740.dmp","upAndRunning":true,"exitStatus":{"status":"ABORTED","signal":-1073741046}}
Opened log file 'C:\vm03-windows\oskar\work\tmp\arangosh_a07908\tmp-6424-1554708290'
Child-SP          RetAddr           Call Site
0000008a`17cfd2c0 00007ff6`a24b933f arangod!abort(void)+0x35 [minkernel\crts\ucrt\src\appcrt\startup\abort.cpp @ 77]
0000008a`17cfd2f0 00007ff6`a24b2551 arangod!arangodb::consensus::State::persist(unsigned int64 index = 0, unsigned int64 term = 0, unsigned int64 millis = 0x00000222`e13d5b30, class arangodb::velocypack::Slice * entry = 0x0000008a`17cfd630, class std::basic_string<char,std::char_traits<char>,std::allocator<char> > * clientId = 0x0000008a`17cfd688 "")+0x43f [c:\vm03-windows\oskar\work\arangodb\arangod\agency\state.cpp @ 106]
0000008a`17cfd600 00007ff6`a2499415 arangod!arangodb::consensus::State::loadCollections(struct TRI_vocbase_t * vocbase = <Value unavailable error>, class arangodb::aql::QueryRegistry * queryRegistry = <Value unavailable error>, bool waitForSync = <Value unavailable error>)+0x191 [c:\vm03-windows\oskar\work\arangodb\arangod\agency\state.cpp @ 781]
0000008a`17cfd720 00007ff6`a3007728 arangod!arangodb::consensus::Agent::load(void)+0x3c5 [c:\vm03-windows\oskar\work\arangodb\arangod\agency\agent.cpp @ 840]
0000008a`17cfdda0 00007ff6`a3005fcf arangod!arangodb::application_features::ApplicationServer::start(void)+0x308 [c:\vm03-windows\oskar\work\arangodb\lib\applicationfeatures\applicationserver.cpp @ 606]
0000008a`17cfec40 00007ff6`a1993ad8 arangod!arangodb::application_features::ApplicationServer::run(int argc = 0n51, char ** argv = 0x00000222`dfbd1620)+0x21f [c:\vm03-windows\oskar\work\arangodb\lib\applicationfeatures\applicationserver.cpp @ 200]
0000008a`17cfeda0 00007ff6`a1995649 arangod!runServer(int argc = 0n51, char ** argv = 0x00000222`dfbd1620, class arangodb::ArangoGlobalContext * context = 0x0000008a`17cff7b0)+0xaf8 [c:\vm03-windows\oskar\work\arangodb\arangod\restserver\arangod.cpp @ 277]
0000008a`17cff6e0 00007ff6`a38ccb9c arangod!main(int argc = 0n51, char ** argv = 0x00000222`dfbd1620)+0x1b9 [c:\vm03-windows\oskar\work\arangodb\arangod\restserver\arangod.cpp @ 412]
(Inline Function) --------`-------- arangod!invoke_main+0x22 [d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 78]
0000008a`17cff840 00007ffb`23f47974 arangod!__scrt_common_main_seh(void)+0x10c [d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 288]
0000008a`17cff880 00007ffb`255ca271 kernel32!BaseThreadInitThunk+0x14
0000008a`17cff8b0 00000000`00000000 ntdll!RtlUserThreadStart+0x21

.  0  Id: 29f4.100c Suspend: 0 Teb: 0000008a`17b95000 Unfrozen
RetAddr           : Args to Child                                                           : Call Site
00007ff6`a24b933f : 00000000`00000001 0000008a`17cfd3f0 00007ff6`a46e964c 0000008a`00000001 : arangod!abort+0x35 [minkernel\crts\ucrt\src\appcrt\startup\abort.cpp @ 77]
00007ff6`a24b2551 : 00000000`00000001 00000222`e13d5b30 00000222`e155ec10 00000222`e13d5b30 : arangod!arangodb::consensus::State::persist+0x43f [c:\vm03-windows\oskar\work\arangodb\arangod\agency\state.cpp @ 106]
00007ff6`a2499415 : 00000000`00000000 00000000`00000000 0000008a`17cfd820 00000000`00000000 : arangod!arangodb::consensus::State::loadCollections+0x191 [c:\vm03-windows\oskar\work\arangodb\arangod\agency\state.cpp @ 781]
00007ff6`a3007728 : 00007ff6`00000000 00000222`e15c84b0 00007ff6`a495fb68 00000000`00000000 : arangod!arangodb::consensus::Agent::load+0x3c5 [c:\vm03-windows\oskar\work\arangodb\arangod\agency\agent.cpp @ 840]
00007ff6`a3005fcf : 0000008a`17cff350 00000000`00000000 00000000`00000000 0000008a`17cff350 : arangod!arangodb::application_features::ApplicationServer::start+0x308 [c:\vm03-windows\oskar\work\arangodb\lib\applicationfeatures\applicationserver.cpp @ 606]
00007ff6`a1993ad8 : 00000222`00000000 00000222`e0353180 00000000`00000000 0000008a`17cfedb0 : arangod!arangodb::application_features::ApplicationServer::run+0x21f [c:\vm03-windows\oskar\work\arangodb\lib\applicationfeatures\applicationserver.cpp @ 200]
00007ff6`a1995649 : 00000000`00000000 00000222`dfbd1620 00000222`dfbd1620 00000000`00000000 : arangod!runServer+0xaf8 [c:\vm03-windows\oskar\work\arangodb\arangod\restserver\arangod.cpp @ 277]
00007ff6`a38ccb9c : 00000000`00000000 00000000`00000000 00007ff6`a650e094 00000000`00000000 : arangod!main+0x1b9 [c:\vm03-windows\oskar\work\arangodb\arangod\restserver\arangod.cpp @ 412]
(Inline Function) : --------`-------- --------`-------- --------`-------- --------`-------- : arangod!invoke_main+0x22 [d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 78]
00007ffb`23f47974 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : arangod!__scrt_common_main_seh+0x10c [d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 288]
00007ffb`255ca271 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : kernel32!BaseThreadInitThunk+0x14
00000000`00000000 : 00000000`00000000 00000000`00000000 00000000`00000000 00000000`00000000 : ntdll!RtlUserThreadStart+0x21
```
The following assert in State.cpp is triggered
```
 101 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE // we should never get here
 102   if (index == 1 && entry.length() != 1 && !entry.hasKey(RECONFIGURE)) {
 103     LOG_TOPIC("cba44", FATAL, Logger::AGENCY)
 104       << "RAFT index 1 must be an holding only a '.agency' entry, however       we got " << entry.toJson();
 105   }
 106   TRI_ASSERT(false);
 107 #endif
```

**Comment by kvahed** [comment]
https://jenkins01.arangodb.biz/job/arangodb-matrix-pr-mac-onlycatch/6243/

**Pull request merged**